### DISCUSSION TITLE
Struggling with Service Mocking: Functional Testing in Symfony

### DISCUSSION THREAD

**Discussion by berkut1** [symfony/symfony#54915] [Q&A] [answered]
Hello!
I embarked on the idea of writing functional (WebTestCase) tests (although, in retrospect, it seems foolish) and encountered some "difficulties" because almost every entity is associated with an external request. How do I properly mock services so that it doesn't become cumbersome? 
For example, there's an entity where a specific check takes place:
```php
public function __construct(
        EnterpriseDispatcherServiceInterface $service, string $name, string $url, string $login, string $password, bool $enabled = true
    )
    {
        $this->name = $name;
        $this->url = $url;
        $this->login = $login;
        $this->password = $password;
        $this->solidcpLoginId = $service->getEnterpriseDispatcherRealUserId($url, $login, $password);
        $this->enabled = $enabled;
        $this->solidcpServers = new ArrayCollection();
        $this->recordEvent(new Event\EnterpriseDispatcherCreated($this));
    }
```
It's a service itself that communicates with another server: 
```php
final readonly class EnterpriseDispatcherService implements EnterpriseDispatcherServiceInterface
{
    public function __construct(
        private EsUsers $esUsers
    ) {}

    public function getEnterpriseDispatcherRealUserId(string $url, string $login, string $password): int
    {
        $this->esUsers->initManual($url, $login, $password);
        try {
            $result = $this->esUsers->getUserByUsername($login);
        } catch (\Exception $e) {
            throw new \DomainException("Soap execution error (Code: {$e->getCode()}, Message: {$e->getMessage()})", $e->getCode(), $e);
        }
        if ($result['IsPeer']) {
            throw new \DomainException("This Login {$login} is Peer. Please use a real User, not Peer");
        }
        return (int)$result['UserId'];
    }
}
```
To mock it without Symfony complaining:
`Symfony\Component\DependencyInjection\Exception\InvalidArgumentException: The "App\Service\Service" service is already initialized, you cannot replace it.`
I have to do:
```php
final class EnterpriseDispatcherServiceMock implements EnterpriseDispatcherServiceInterface
{
    public function getEnterpriseDispatcherRealUserId(string $url, string $login, string $password): int
    {
        return 1001;
    }
}
```
In some of these mocks, I add the ability to dynamically substitute values to simulate different responses from the external service. Then, I add the corresponding entry in `services_test.yaml`.

```yaml
services:
  _defaults:
    autowire: true

  App\Service\EnterpriseDispatcherService:
    public: true
    class: App\Tests\Mock\EnterpriseDispatcherServiceMock
```

As a result, the code grows catastrophically. If I need to test 20 services, I have to create 20 interfaces, 20 mock classes, and specify 20 entries in `services_test.yaml`. 
Is there a simpler solution, or are functional tests a rabbit hole better left untouched?"

UPD: Symfony v6.4.7

**Comment by Trismegiste** [comment]
> As a result, the code grows catastrophically

It's because your code does not follow SOLID, therefore difficult to mock. In the constructor :
* You get something instead of injecting it (DIP violation a.k.a 'Hollywood principle')
* You have a new instead of injecting a factory (SRP and OCP violations)

```
$this->solidcpLoginId = $service->getEnterpriseDispatcherRealUserId($url, $login, $password);
// ...
$this->recordEvent(new Event\EnterpriseDispatcherCreated($this));
```

More on https://en.wikipedia.org/wiki/SOLID

HTH

**Reply by berkut1** [reply]
This constructor is from Entity.
```
#[ORM\Entity]
class EnterpriseDispatcher implements AggregateRoot
{
     public function __construct(
        EnterpriseDispatcherServiceInterface $service, string $name, string $url, string $login, string $password, bool $enabled = true
    )
    {....}
}
```

I use `EnterpriseDispatcherServiceInterface` here because I want to prevent any possibility of setting an incorrect value in the `$this->solidcpLoginId` property.
Is this an incredibly stupid solution? Would it be better to simply use a factory where the necessary check can be performed and the entity created?
When I created this entity, my intention was to prevent the possibility of setting incorrect values in any way possible (even if this code will only be used by me:) ), which is why this service was created and uses in constructor. 
Thanks.

**Reply by berkut1** [reply]
Anyway, if I move this check up to the controller, it won't solve the issue I raised. When testing via WebTestCase (https://symfony.com/doc/current/testing.html#write-your-first-application-test), external services will still be triggered, and I need to mock them, but I can't just replace them with `$this->client->getContainer()->set();` because it will throw an error:
`Symfony\Component\DependencyInjection\Exception\InvalidArgumentException: The "App\Service\Service" service is already initialized, you cannot replace it.`

**Reply by pkly** [reply, marked as answer]
I personally find that including any arguments in an entity constructor needlessly complicates everything. You also seem to use a service inside of your entity, which is not great. I would simply suggest creating your entity through a factory instead, which would give you yet another service.

As for mocking, I do something similar in a lot of our tests and I generally find that booting the container, and only booting it, should allow you to replace services. If a service cannot be replaced it means it was already used in something when the container started, which could've happened if you say, made a request in the test case.

Example from one of our WebTestCase tests below.

```php
        parent::setUp();
        self::createClient();
        self::getContainer()->set(FooHandler::class, $this->handler = $this->createMock(FooHandler::class));
```

**Reply by berkut1** [reply]
@pkly Thanks. I will move that service to a factory:)

Now about the issue.
Your solution help only in half. I don't get anymore:
`Symfony\Component\DependencyInjection\Exception\InvalidArgumentException: The "App\Service\Service" service is already initialized, you cannot replace it.`

But also it doesn't use the mocked service.
```php
public function testCreate(): void
    {
        $client = self::createClient();

        $service = $this->createMock(EnterpriseDispatcherService::class);
        $service->expects($this->once())
            ->method('getEnterpriseDispatcherRealUserId')
            ->willReturn(12333);
        self::getContainer()->set(EnterpriseDispatcherService::class, $service);

        $this->loginAs('test_admin');
        $client->request('GET', '/enterprise-dispatchers/create');

        $crawler = $client->submitForm('Create', [
            'form[name]' => $name = 'Test Enterprise Dispatcher',
            'form[url]' => 'http://127.0.0.2:9003',
            'form[login]' => 'test_login',
            'form[password]' => 'test_password',
        ]);	
    }
```
Here, after submitting the form `$client->submitForm` where the service should be triggered, it will not return 12333; it will still attempt to use the real service instead of the mocked one.

Where Am I wrong?

P.S Symfony v6.4.7

**Reply by berkut1** [reply]
Well, my last issue was resolved by adding `$client->disableReboot();`. 

Thanks to @pkly for pointing the way :)

**Reply by pkly** [reply]
Glad I could help, web cases are a bit trickier to modify because of the container resets between requests.

**Reply by berkut1** [reply]
@pkly Thanks. I have the last question :) It's about a service in an Entity.
In the book Domain-Driven Design (Eric Evans), page 76 https://fabiofumarola.github.io/nosql/readingMaterial/Evans03.pdf
there is a quote:

> When a significant process or transformation in the domain is not a natural responsibility of an ENTITY or VALUE OBJECT, add an operation to the model as standalone interface declared as a SERVICE. Define the interface in terms of the language of the model and make sure the operation name is part of the UBIQUITOUS LANGUAGE. Make the SERVICE stateless.

As I understand I can inject services as I do in my example here.
Am I understand it incorrectly? Sorry my English isn't main language and some word can be tricky to understand..

**Reply by MatTheCat** [reply]
The method of injecting services into entities is called “Disconnected Domain Model”, but this is not what the quote is about: it tells you that if a change of the domain’s state is not the responsibility of an entity/value object, then it is a service’s.

In your case, the corresponding service is the factory @pkly proposed.

By the way, you really should rename `EnterpriseDispatcherService` because it hides what it does (if it does many things, this is probably a violation of the Single Responsibility Principle and you should have many services instead). An `EnterpriseDispatcher` factory would be called `EnterpriseDispatcherFactory` e.g.

**Reply by berkut1** [reply]
@MatTheCat thanks for the explanation!
